the start of an implementation of files.block

morrison12 opened the pull request:
This is the beginning of an implementation of `files.block` (and `files.Block`) as contemplated in #565.

The main missing feature I can see is `assume_file_present` which is need for the common use case of installing a package and then configuring it although there could well be others.

Some "notes:
1. The current implementation uses `awk` which I assumed was a reasonably dependency as even busybody seems to have it ?
1. Not all `awk` implementations support in place editing so the code currently does the "write to a temporary file, if things worked then adjust the user/group/mode and mv it to the desired location" jig...  
1. I've tested this with a few `awk` implementations (both the native and GNU awks on macOS, GNU awk on Fedora and busybox awk in OpenWrt)

Some questions:
1. Given a file is created in the case where it doesn't exist, should I allow the caller to supplier the `user`, `group` and `mode` ?
1.  Despite the temporary file being used, for consistency with other operations would a `backup` option be useful ?
2. I have some, admittedly not speedy, functional tests; is there somewhere they should be stored ?

Commit:
feature: enhance files.block to augment regexes in the same way as files.line

Commit:
fix: update files typestubs for last change to API

Fizzadar commented on pyinfra/operations/files.py:
```diff
@@ -320,17 +323,17 @@ def line(
-        match_line = "^.*{0}".format(match_line)
-    if not match_line.endswith("$"):
-        match_line = "{0}.*$".format(match_line)
+    match_line = adjust_regex(line, escape_regex_characters)
```
Nice!

Fizzadar commented on pyinfra/operations/files.py:
```diff
@@ -1526,3 +1529,189 @@ def directory(
+
+        # add entry to /etc/host
+        files.marked_block(
+            name = "add IP address for red server",
```
For consistency these should not have spaces between the keyword & value.

morrison12 replied:
Fixed.

Fizzadar commented on pyinfra/operations/files.py:
```diff
@@ -1526,3 +1529,189 @@ def directory(
+            redirect = ">" if (current is None) else ">>"
+            stdin = "- " if ((current == []) and before) else ""
+            # here = hex(random.randint(0, 2147483647))
+            here = "HERE"
```
Should we use something like `PYINFRAHERE` to avoid clashes with any user content in the block?

morrison12 replied:
I changed it.  Originally I had a random number but that broke the idempotence testing.

Fizzadar reviewed (commented):
THANK YOU! @morrison12 this has been in the backlog way too long and I've been unable to un-wrangle how to approach it, massive thank you for figuring it out!

PR looks great overall, few very minor comments. Addressing your questions below:

> The main missing feature I can see is assume_file_present which is need for the common use case of installing a package and then configuring it although there could well be others.

Ah yes, very unfortunate gotcha. This operation seems like it'd be particularly complex to support. Maybe it could just append the block, or maybe we don't support it? Happy to leave this out of scope for this PR entirely.

 > The current implementation uses awk which I assumed was a reasonably dependency as even busybody seems to have it ?
 > Not all awk implementations support in place editing so the code currently does the "write to a temporary file, if things worked then adjust the user/group/mode and mv it to the desired location" jig...

👍 


Some questions:

> Given a file is created in the case where it doesn't exist, should I allow the caller to supplier the user, group and mode ?

Good question, I would say no for now given that `files.line` is the same. If needed this can be combined with `files.file` to achieve this.

> Despite the temporary file being used, for consistency with other operations would a backup option be useful ?

I think so for keeping the old version around somewhere.

> I have some, admittedly not speedy, functional tests; is there somewhere they should be stored ?

What do they look like? :) Maybe under `tests/test_operations/test_files.py`?

Commit:
make heredoc marker more distinct and clean up comments

morrison12 commented:
> > The main missing feature I can see is assume_file_present which is need for the common use case of installing a package and then configuring it although there could well be others.
> 
> Ah yes, very unfortunate gotcha. This operation seems like it'd be particularly complex to support. Maybe it could just append the block, or maybe we don't support it? Happy to leave this out of scope for this PR entirely.
> 
I'll happily file the bug report once this is released and work on a solution but need to find some time to make sure it covers all the cases.  Perhaps mark the operation as "beta" in the meantime ?

Commit:
enh: add backup option to files.block operation

Commit:
fix: correct type stub for file.block operation

Commit:
formatting update to files typestub

Commit:
Revert "formatting update to files typestub"

This reverts commit 0dd2d742dad8bfca3b9bb0e57180146054abe9c7.

Commit:
fix: correct type stub for file.block operation

Commit:
another attempt at the typestub for files after adding files.block

Fizzadar committed:
feature: add/remove blocks of text (surrounded by markers) from a file

# Conflicts:
#	pyinfra/facts/files.py
#	pyinfra/operations/files.py
#	pyinfra/operations/files.pyi

Fizzadar committed:
Add log warning to indicate `files.block` is a beta operation

Fizzadar reviewed (approved):
Thank you @morrison12 for implementing this, fantastic as ever! I have rebased & added the beta warning to it ready to release :)

morrison12 commented:
Thanks @Fizzadar.  I've opened #900 to track the need for `assume_file_present`.
